src/analysis/inner_product_space/pi_L2.lean

```diff
@@ -381,3 +381,86 @@ def linear_isometry_equiv.from_orthogonal_span_singleton
   (n : ℕ) [fact (finrank 𝕜 E = n + 1)] {v : E} (hv : v ≠ 0) :
   (𝕜 ∙ v)ᗮ ≃ₗᵢ[𝕜] (euclidean_space 𝕜 (fin n)) :=
 linear_isometry_equiv.of_inner_product_space (finrank_orthogonal_span_singleton hv)
+
+section complex
+
+variables {V : Type*}
+[inner_product_space 𝕜 V]
+[finite_dimensional 𝕜 V]
+
+variables {S : submodule 𝕜 V} {L : S →ₗᵢ[𝕜] V}
+
+instance S_complete : complete_space S := finite_dimensional.complete 𝕜 S
+
+instance V_complete : complete_space V := finite_dimensional.complete 𝕜 V
+
+open finite_dimensional
+
+/-- Let `S` be a subspace of a finite-dimensional complex inner product space `V`.  A linear
+isometry mapping `S` into `V` can be extended to a full isometry of `V`. -/
+noncomputable def linear_isometry.extend (L : S →ₗᵢ[𝕜] V): V →ₗᵢ[𝕜] V :=
+begin
+  -- Build an isometry from Sᗮ to L(S)ᗮ through euclidean_space
+  let d := finrank 𝕜 Sᗮ,
+  have dim_S : finrank 𝕜 Sᗮ = d := rfl,
+  let LS := L.to_linear_map.range,
+  have dim_LS_perp : finrank 𝕜 LSᗮ = d,
+  calc  finrank 𝕜 LSᗮ = finrank 𝕜 V - finrank 𝕜 LS : by simp only
+      [← LS.finrank_add_finrank_orthogonal, add_tsub_cancel_left]
+    ...               = finrank 𝕜 V - finrank 𝕜 S : by simp only
+      [linear_map.finrank_range_of_inj L.injective]
+    ...               = finrank 𝕜 Sᗮ : by simp only
+      [← S.finrank_add_finrank_orthogonal, add_tsub_cancel_left]
+    ...               = d : dim_S,
+  let L1 := ((fin_std_orthonormal_basis dim_S).to_orthonormal_basis
+    (fin_std_orthonormal_basis_orthonormal dim_S)).repr.to_linear_isometry,
+  let L2 := ((fin_std_orthonormal_basis dim_LS_perp).to_orthonormal_basis
+    (fin_std_orthonormal_basis_orthonormal dim_LS_perp)).repr.symm.to_linear_isometry,
+  let L3 := (LS)ᗮ.subtypeₗᵢ.comp (L2.comp L1),
```

Comment:
I think this is a little clearer as:
```lean
  let E : Sᗮ ≃ₗᵢ[𝕜] LSᗮ,
  { have dim_LS_perp : finrank 𝕜 LSᗮ = d,
    calc  finrank 𝕜 LSᗮ = finrank 𝕜 V - finrank 𝕜 LS : by simp only
        [← LS.finrank_add_finrank_orthogonal, add_tsub_cancel_left]
      ...               = finrank 𝕜 V - finrank 𝕜 S : by simp only
        [linear_map.finrank_range_of_inj L.injective]
      ...               = finrank 𝕜 Sᗮ : by simp only
        [← S.finrank_add_finrank_orthogonal, add_tsub_cancel_left]
      ...               = d : dim_S,
    let BS := ((fin_std_orthonormal_basis dim_S).to_orthonormal_basis
      (fin_std_orthonormal_basis_orthonormal dim_S)),
    let BLS := ((fin_std_orthonormal_basis dim_LS_perp).to_orthonormal_basis
      (fin_std_orthonormal_basis_orthonormal dim_LS_perp)),
    exact BS.repr.trans BLS.repr.symm },
  let L3 := (LS)ᗮ.subtypeₗᵢ.comp E.to_linear_isometry,
```
as this encapsulates the fact that this is all a game to find that equiv.

Reply:
I like this change, and I had not thought to use `have` in this way.